Add host-side Triton TMA support to Dynamo

aakhundov opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #137950
* __->__ #137677

This adds Dynamo tracing support for the host-side Triton TMA API (see `create_2d_tma_descriptor` calls on the host in the [Triton tutorial](https://triton-lang.org/main/getting-started/tutorials/09-persistent-matmul.html#sphx-glr-getting-started-tutorials-09-persistent-matmul-py)). A few notes:

- Here we assume the availability of the host-side TMA API added to upstream Triton in https://github.com/triton-lang/triton/pull/4498. As of time of writing, this is not a part of the PT2 OSS Triton pin (although back-ported internally). OSS Triton pin update should be done in December 2024.
- To capture the chain of calls `t.data_ptr() --> create_{1d,2d}_tma_descriptor(ptr, ...) --> kernel[grid](tma_desc, ...)`, we add three new variable trackers: `DataPtrVariable`, `CreateTMADescriptorVariable` (for the function), `TMADescriptorVariable` (for TMA descriptor object). This is to maintain the path back from the Triton kernel to the Tensor from which the TMA descriptor has been created.
- The newly introduced variables have `reconstruct` methods used in case of graph breaks.
- The `tma_descriptor_metadata` extracted from the captured `create_{1d,2d}_tma_descriptor` calls is propagated through the HOPs in Dynamo and AOTAutograd to be used by the downstream compiler (e.g., Inductor). See the unit tests for how the captured HOP arguments look like.
- In the Dynamo-captured fx graph, we replace the TMA descriptor arguments of the Triton kernel by the underlying Tensors, to be able to track the input/output relationships in terms of Tensors.
- In the Triton kernel mutation analysis pass (in AOTAutograd), we use the `tt.experimental_descriptor_store` TTIR op to detect mutations of the underlying tensors via TMA descriptors. So that downstream AOTAutograd can perform functionalizations as required.
- JIT Inductor and AOT Inductor support will be implemented in follow-up PRs.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang @rec

Differential Revision: [D64404928](https://our.internmc.facebook.com/intern/diff/D64404928)

aakhundov committed:
Add host-side Triton TMA support to Dynamo

[ghstack-poisoned]

aakhundov committed:
Update on "Add host-side Triton TMA support to Dynamo"


This adds Dynamo tracing support to the host-side Triton TMA API (see `create_2d_tma_descriptor` calls on the host in the [Triton tutorial](https://triton-lang.org/main/getting-started/tutorials/09-persistent-matmul.html#sphx-glr-getting-started-tutorials-09-persistent-matmul-py)).

Design notes TBA.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

[ghstack-poisoned]

aakhundov committed:
Update on "Add host-side Triton TMA support to Dynamo"


This adds Dynamo tracing support to the host-side Triton TMA API (see `create_2d_tma_descriptor` calls on the host in the [Triton tutorial](https://triton-lang.org/main/getting-started/tutorials/09-persistent-matmul.html#sphx-glr-getting-started-tutorials-09-persistent-matmul-py)).

Design notes TBA.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

[ghstack-poisoned]

aakhundov committed:
Update on "Add host-side Triton TMA support to Dynamo"


This adds Dynamo tracing support to the host-side Triton TMA API (see `create_2d_tma_descriptor` calls on the host in the [Triton tutorial](https://triton-lang.org/main/getting-started/tutorials/09-persistent-matmul.html#sphx-glr-getting-started-tutorials-09-persistent-matmul-py)).

Design notes TBA.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

[ghstack-poisoned]

zou3519 commented on torch/_dynamo/variables/tensor.py:
```diff
@@ -1448,3 +1448,13 @@ def reconstruct(self, codegen):
         codegen.call_method(0)
+
+
+class DataPtrVariable(VariableTracker):
+    def __init__(
+        self,
+        tensor: TensorVariable,
+        **kwargs,
+    ) -> None:
+        super().__init__(**kwargs),
+        self.tensor = tensor
```
Does this really work without no reconstruct method? :O

aakhundov replied:
I haven't added graph break tests yet, will do as we've discussed the details of this offline.

aakhundov replied:
@zou3519 @anijain2305 I've added `reconstruct` methods to the `DataPtrVariable` and `TMADescriptorVariable` classes. Also added unit tests with graph breaks: both for the TMA in Triton kernels and `data_ptr + data_ptr` case. Please let me know if this looks sufficient.

zou3519 commented on test/inductor/test_triton_kernels.py:
```diff
@@ -1630,6 +1636,158 @@ def f(x, y, z):
+        # reset logging state
+        torch._logging.set_logs()
+
+        output = "\n".join(record.getMessage() for record in log.records)
+        output = re.sub(r"\033\[[0-9;]*m", "", output)  # remove coloring
+
+        # structured capturing of the TMA descriptor metadata
+        self.assertIn(
+            "tma_descriptor_metadata = {'in_desc_ptr0': ([301], [256], 4), 'in_desc_ptr1': ([301], [256], 4), 'out_desc_ptr': ([301], [256], 4)}",
+            output,
+        )
+        # functionalization: the output should be marked as mutated
+        self.assertIn("tensors_to_clone = ['out_desc_ptr']", output)
```
is it possible to just do an assertExpectedInline? This makes it clearer what's going on overall with the design.

For example: "out_desc_ptr" is a key in both the tma_descriptor_metadata and the kwargs dicts passed to triton_kernel_wrap. I learned this the other day when you sent me some graphs; putting the graphs here in code would help other people figure out what's going on

zou3519 commented on test/inductor/test_triton_kernels.py:
```diff
@@ -1630,6 +1636,158 @@ def f(x, y, z):
+            dynamic=dynamic,
+        )(a, b)
+
+        torch.testing.assert_close(eager_out, expected_out)
+        torch.testing.assert_close(compiled_out, expected_out)
```
Use self.assertEqual over torch.testing.assert_close. @kit1980 this would be another good lint rule

zou3519 commented on torch/_dynamo/variables/functions.py:
```diff
@@ -1108,6 +1111,18 @@ def call_HOP(self, variable, grids, combined_args_raw, tx):
+        # instances) with the underlying tensors, while moving the
+        # TMA descriptor-related metadata to a separate argument,
+        # so that we can reconstruct the TMA descriptors downstream
+        tma_descriptor_metadata: TMADescriptorMetadata = {}
```
type hint wrong?

aakhundov replied:
Not sure what you mean, could you elaborate? `TMADescriptorMetadata` is defined as

```
TMADescriptorMetadata = Dict[
    str,  # kernel parameter name
    Tuple[
        List[Union[int, SymInt]],  # dims
        List[Union[int, SymInt]],  # block_dims
        Union[int, SymInt],  # element_size
    ],
]
```

and, in the few lines below, `tma_descriptor_metadata` is filled in accordingly?

zou3519 replied:
Sorry, I didn't realize it was a dict

zou3519 reviewed (commented):
Looks good overall, let's add some more testing around graph breaks.

mobicham commented:
Is this gonna support pre-allocated descriptors on the host-side ? Creating descriptors on the fly for each kernel launch is very slow as you probably know. 
Thanks!

aakhundov commented:
> Is this gonna support pre-allocated descriptors on the host-side ? Creating descriptors on the fly for each kernel launch is very slow as you probably know. Thanks!

@mobicham thanks for chiming in! Yes, we want to support host-side TMA descriptor preallocation (and filling) with this PR and a (soon-to-be) follow up adding Inductor support. Along the lines of the `matmul_tma_persistent` in the [Triton tutorial #9](https://triton-lang.org/main/getting-started/tutorials/09-persistent-matmul.html). Is this what you're referring to?

mobicham commented:
@aakhundov perfect, that's exactly what I am looking for, thank you!

aakhundov committed:
Update on "Add host-side Triton TMA support to Dynamo"


This adds Dynamo tracing support for the host-side Triton TMA API (see `create_2d_tma_descriptor` calls on the host in the [Triton tutorial](https://triton-lang.org/main/getting-started/tutorials/09-persistent-matmul.html#sphx-glr-getting-started-tutorials-09-persistent-matmul-py)).

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

[ghstack-poisoned]

aakhundov committed:
Update on "Add host-side Triton TMA support to Dynamo"


This adds Dynamo tracing support for the host-side Triton TMA API (see `create_2d_tma_descriptor` calls on the host in the [Triton tutorial](https://triton-lang.org/main/getting-started/tutorials/09-persistent-matmul.html#sphx-glr-getting-started-tutorials-09-persistent-matmul-py)).

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

[ghstack-poisoned]

aakhundov committed:
Update on "Add host-side Triton TMA support to Dynamo"


This adds Dynamo tracing support for the host-side Triton TMA API (see `create_2d_tma_descriptor` calls on the host in the [Triton tutorial](https://triton-lang.org/main/getting-started/tutorials/09-persistent-matmul.html#sphx-glr-getting-started-tutorials-09-persistent-matmul-py)). A few notes:

- Here we assume the availability of the host-side TMA API added to upstream Triton in https://github.com/triton-lang/triton/pull/4498. As of time of writing, this is not a part of the PT2 OSS Triton pin (although back-ported internally). OSS Triton pin update should be done in December 2024.
- To capture the chain of calls `t.data_ptr() --> create_{1d,2d}_tma_descriptor(ptr, ...) --> kernel[grid](tma_desc, ...)`, we add three new variable trackers: `DataPtrVariable`, `CreateTMADescriptorVariable` (for the function), `TMADescriptorVariable` (for TMA descriptor object). This is to maintain the path back from the Triton kernel to the Tensor from which the TMA descriptor has been created.
- The newly introduced variables have `reconstruct` methods used in case of graph breaks.
- The `tma_descriptor_metadata` extracted from the captured `create_{1d,2d}_tma_descriptor` calls is propagated through the HOPs in Dynamo and AOTAutograd to be used by the downstream compiler (e.g., Inductor). See the unit tests for how the captured HOP arguments look like.
- In the Dynamo-captured fx graph, we replace the TMA descriptor arguments of the Triton kernel by the underlying Tensors, to be able to track the input/output relationships in terms of Tensors.
- In the Triton kernel mutation analysis pass (in AOTAutograd), we use the `tt.experimental_descriptor_store` TTIR op to detect mutations of the underlying tensors via TMA descriptors. So that downstream AOTAutograd can perform functionalizations as required.
- JIT Inductor and AOT Inductor support will be implemented in follow-up PRs.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

[ghstack-poisoned]

aakhundov committed:
Update on "Add host-side Triton TMA support to Dynamo"


This adds Dynamo tracing support for the host-side Triton TMA API (see `create_2d_tma_descriptor` calls on the host in the [Triton tutorial](https://triton-lang.org/main/getting-started/tutorials/09-persistent-matmul.html#sphx-glr-getting-started-tutorials-09-persistent-matmul-py)). A few notes:

- Here we assume the availability of the host-side TMA API added to upstream Triton in https://github.com/triton-lang/triton/pull/4498. As of time of writing, this is not a part of the PT2 OSS Triton pin (although back-ported internally). OSS Triton pin update should be done in December 2024.
- To capture the chain of calls `t.data_ptr() --> create_{1d,2d}_tma_descriptor(ptr, ...) --> kernel[grid](tma_desc, ...)`, we add three new variable trackers: `DataPtrVariable`, `CreateTMADescriptorVariable` (for the function), `TMADescriptorVariable` (for TMA descriptor object). This is to maintain the path back from the Triton kernel to the Tensor from which the TMA descriptor has been created.
- The newly introduced variables have `reconstruct` methods used in case of graph breaks.
- The `tma_descriptor_metadata` extracted from the captured `create_{1d,2d}_tma_descriptor` calls is propagated through the HOPs in Dynamo and AOTAutograd to be used by the downstream compiler (e.g., Inductor). See the unit tests for how the captured HOP arguments look like.
- In the Dynamo-captured fx graph, we replace the TMA descriptor arguments of the Triton kernel by the underlying Tensors, to be able to track the input/output relationships in terms of Tensors.
- In the Triton kernel mutation analysis pass (in AOTAutograd), we use the `tt.experimental_descriptor_store` TTIR op to detect mutations of the underlying tensors via TMA descriptors. So that downstream AOTAutograd can perform functionalizations as required.
- JIT Inductor and AOT Inductor support will be implemented in follow-up PRs.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

[ghstack-poisoned]

aakhundov committed:
Update on "Add host-side Triton TMA support to Dynamo"


This adds Dynamo tracing support for the host-side Triton TMA API (see `create_2d_tma_descriptor` calls on the host in the [Triton tutorial](https://triton-lang.org/main/getting-started/tutorials/09-persistent-matmul.html#sphx-glr-getting-started-tutorials-09-persistent-matmul-py)). A few notes:

- Here we assume the availability of the host-side TMA API added to upstream Triton in https://github.com/triton-lang/triton/pull/4498. As of time of writing, this is not a part of the PT2 OSS Triton pin (although back-ported internally). OSS Triton pin update should be done in December 2024.
- To capture the chain of calls `t.data_ptr() --> create_{1d,2d}_tma_descriptor(ptr, ...) --> kernel[grid](tma_desc, ...)`, we add three new variable trackers: `DataPtrVariable`, `CreateTMADescriptorVariable` (for the function), `TMADescriptorVariable` (for TMA descriptor object). This is to maintain the path back from the Triton kernel to the Tensor from which the TMA descriptor has been created.
- The newly introduced variables have `reconstruct` methods used in case of graph breaks.
- The `tma_descriptor_metadata` extracted from the captured `create_{1d,2d}_tma_descriptor` calls is propagated through the HOPs in Dynamo and AOTAutograd to be used by the downstream compiler (e.g., Inductor). See the unit tests for how the captured HOP arguments look like.
- In the Dynamo-captured fx graph, we replace the TMA descriptor arguments of the Triton kernel by the underlying Tensors, to be able to track the input/output relationships in terms of Tensors.
- In the Triton kernel mutation analysis pass (in AOTAutograd), we use the `tt.experimental_descriptor_store` TTIR op to detect mutations of the underlying tensors via TMA descriptors. So that downstream AOTAutograd can perform functionalizations as required.
- JIT Inductor and AOT Inductor support will be implemented in follow-up PRs.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

[ghstack-poisoned]

anijain2305 reviewed (commented):
Dynamo parts LGTM.

aakhundov committed:
Update on "Add host-side Triton TMA support to Dynamo"


This adds Dynamo tracing support for the host-side Triton TMA API (see `create_2d_tma_descriptor` calls on the host in the [Triton tutorial](https://triton-lang.org/main/getting-started/tutorials/09-persistent-matmul.html#sphx-glr-getting-started-tutorials-09-persistent-matmul-py)). A few notes:

- Here we assume the availability of the host-side TMA API added to upstream Triton in https://github.com/triton-lang/triton/pull/4498. As of time of writing, this is not a part of the PT2 OSS Triton pin (although back-ported internally). OSS Triton pin update should be done in December 2024.
- To capture the chain of calls `t.data_ptr() --> create_{1d,2d}_tma_descriptor(ptr, ...) --> kernel[grid](tma_desc, ...)`, we add three new variable trackers: `DataPtrVariable`, `CreateTMADescriptorVariable` (for the function), `TMADescriptorVariable` (for TMA descriptor object). This is to maintain the path back from the Triton kernel to the Tensor from which the TMA descriptor has been created.
- The newly introduced variables have `reconstruct` methods used in case of graph breaks.
- The `tma_descriptor_metadata` extracted from the captured `create_{1d,2d}_tma_descriptor` calls is propagated through the HOPs in Dynamo and AOTAutograd to be used by the downstream compiler (e.g., Inductor). See the unit tests for how the captured HOP arguments look like.
- In the Dynamo-captured fx graph, we replace the TMA descriptor arguments of the Triton kernel by the underlying Tensors, to be able to track the input/output relationships in terms of Tensors.
- In the Triton kernel mutation analysis pass (in AOTAutograd), we use the `tt.experimental_descriptor_store` TTIR op to detect mutations of the underlying tensors via TMA descriptors. So that downstream AOTAutograd can perform functionalizations as required.
- JIT Inductor and AOT Inductor support will be implemented in follow-up PRs.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang rec

[ghstack-poisoned]

aakhundov commented on torch/_dynamo/variables/tensor.py:
```diff
@@ -1439,3 +1439,18 @@ def reconstruct(self, codegen):
+        self.from_tensor = from_tensor
+
+    def reconstruct(self, codegen):
+        codegen(self.from_tensor, allow_cache=False)
```
@anijain2305 @zou3519 without `allow_cache=False` here, the `TestAutograd.test_no_grad_copy` is failing with Dynamo under Python 3.12.

zou3519 replied:
I don't know what allow_cache is, @anijain2305 does this look concerning?

aakhundov commented:
@zou3519 I believe, all comments are addressed. Could you take another look? Thanks!

zou3519 commented on test/dynamo/test_misc.py:
```diff
@@ -11534,6 +11534,19 @@ def get_rng():
+
+        expected = f(a, b)
+        actual = torch.compile(f, backend="eager")(a, b)
+
```
Can we do an assertion that there was a graph break?

aakhundov replied:
I've added another test (`test_data_ptr_graph_break_aten`) in which I can assert on the graph break. In the original test (now named `test_data_ptr_graph_break_builtin`), `counters["graph_break"]` remains empty, although I can see a graph break in the `TORCH_LOGS="+dynamo"`. Would this state be sufficient for testing?

aakhundov replied:
Yes, with `fullgraph=True`, we get this error:

```
torch._dynamo.exc.Unsupported: builtin: add [<class 'torch._dynamo.variables.tensor.DataPtrVariable'>, <class 'torch._dynamo.variables.tensor.DataPtrVariable'>] False
```

As @williamwen42 has pointed out:

> it looks like only certain classes of graph breaks actually contribute to counters["graph_break"] - graph breaks caused by STORE_ATTRs and function call bytecode. in your graph break case, BINARY_ADD causes the graph break, which is not one of those classes. it seems generally that graph breaks that cause the entire frame to be skipped don't count.

I've added `self.assertRaises` with `fullgraph=True` to the test to make sure there is a graph break. Please let me know if this is good enough.

zou3519 replied:
Weird, thanks for looking into this (and adding that test)

zou3519 commented on test/dynamo/test_misc.py:
```diff
@@ -11534,6 +11534,19 @@ def get_rng():
+    def test_data_ptr_graph_break(self):
+        def f(a, b):
+            # builtin + not implemented for DataPtrVariable
+            return a.data_ptr() + b.data_ptr()
```
Could you also test an aten op, like torch.add(a, a.data_ptr())?

zou3519 commented on torch/_dynamo/testing.py:
```diff
@@ -245,6 +246,27 @@ def __call__(
+            self.graphs.append(gm)
+            return gm.forward
+
+        return aot_eager_decomp_partition(
```
btw, can you modify the `aot_eager` backend instead? We're modifying `aot_eager_decomp_partition`. I know I sent you the link to the source code for aot_eager_decomp_partition, I actually meant to send you the source code for `aot_eager`

zou3519 replied:
NB: backend="aot_eager" is different from backend="aot_eager_decomp_partition" :P

zou3519 commented on torch/_dynamo/variables/functions.py:
```diff
@@ -1154,3 +1177,92 @@ def specialize_symbolic(self, arg: Any) -> Any:
+
+        if not isinstance(ptr, variables.DataPtrVariable):
+            raise Unsupported(
+                "create_{1d,2d}_tma_descriptor can only be traced "
+                "when the upstream .data_ptr() call has been traced."
```
Let's make the graph break error message a bit more actionable. "Please ensure there were no graph breaks between create_{1d,2d}_tma_descriptor and the upstream .data_ptr() call"

zou3519 replied:
Let's make the graph break error message a bit more actionable. "Please ensure there were no graph breaks between create_{1d,2d}_tma_descriptor and the upstream .data_ptr() call"

zou3519 commented on torch/_dynamo/variables/functions.py:
```diff
@@ -1154,3 +1177,92 @@ def specialize_symbolic(self, arg: Any) -> Any:
+        element_size = kwargs["ptr"] if "ptr" in kwargs else args[-1]
+
+        return TMADescriptorVariable(
+            data_ptr=ptr,  # pyre-ignore[6]
+            dims=dims,  # pyre-ignore[6]
+            block_dims=block_dims,  # pyre-ignore[6]
+            element_size=element_size,  # pyre-ignore[6]
```
why the pyre-ignores?

aakhundov replied:
Leftover from the co-dev. Thanks for catching, will remove.

zou3519 commented on test/inductor/test_triton_kernels.py:
```diff
@@ -1630,6 +1635,218 @@ def f(x, y, z):
+    add_2 = arg0_1 + 256
+    sub_1 = add_2 - 1;  add_2 = None
+    floordiv = sub_1 // 256;  sub_1 = None
+    triton_kernel_wrapper_functional_proxy = torch.ops.higher_order.triton_kernel_wrapper_functional(kernel_idx = 0, constant_args_idx = 1, grid = [(floordiv, 1, 1)], tma_descriptor_metadata = {'in_desc_ptr0': ([arg0_1], [256], 4), 'in_desc_ptr1': ([arg0_1], [256], 4), 'out_desc_ptr': ([arg0_1], [256], 4)}, kwargs = {'in_desc_ptr0': arg1_1, 'in_desc_ptr1': arg2_1, 'out_desc_ptr': full}, tensors_t ...
```
Did we expect the [256] to be dynamic?

zou3519 replied:
Hmm looks like no, that's just the BLOCK_SIZE

aakhundov replied:
No, it should be static as `BLOCK_SIZE: tl.constexpr`. `block_dims` are generally constant / fixed for TMA descriptors.

aakhundov committed:
Update

[ghstack-poisoned]

aakhundov committed:
Update

[ghstack-poisoned]

aakhundov committed:
Update

[ghstack-poisoned]

aakhundov committed:
Update

[ghstack-poisoned]

aakhundov commented:
@aakhundov has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D64404928).

aakhundov commented:
@pytorchbot merge
